sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala

```diff
@@ -1262,6 +1275,85 @@ class JDBCV2Suite extends QueryTest with SharedSparkSession with ExplainSuiteHel
     checkAnswer(df17, Seq(Row(6, "jen", 12000, 1200, true)))
   }
 
+  test("SPARK-38432: escape the quote, _ and % for DS V2 pushdown") {
+    val df1 = spark.table("h2.test.address").filter($"email".startsWith("abc_"))
+    checkFiltersRemoved(df1)
+    checkPushedInfo(df1, "PushedFilters: [EMAIL IS NOT NULL, EMAIL LIKE 'abc\\_%']")
+    checkAnswer(df1,
+      Seq(Row("abc_%def@gmail.com"), Row("abc_'%def@gmail.com"), Row("abc_def@gmail.com")))
+
+    val df2 = spark.table("h2.test.address").filter($"email".startsWith("abc%"))
+    checkFiltersRemoved(df2)
+    checkPushedInfo(df2, "PushedFilters: [EMAIL IS NOT NULL, EMAIL LIKE 'abc\\%%']")
+    checkAnswer(df2, Seq(Row("abc%_def@gmail.com"), Row("abc%def@gmail.com")))
+
+    val df3 = spark.table("h2.test.address").filter($"email".startsWith("abc%_"))
+    checkFiltersRemoved(df3)
+    checkPushedInfo(df3, "PushedFilters: [EMAIL IS NOT NULL, EMAIL LIKE 'abc\\%\\_%']")
+    checkAnswer(df3, Seq(Row("abc%_def@gmail.com")))
+
+    val df4 = spark.table("h2.test.address").filter($"email".startsWith("abc_%"))
+    checkFiltersRemoved(df4)
+    checkPushedInfo(df4, "PushedFilters: [EMAIL IS NOT NULL, EMAIL LIKE 'abc\\_\\%%']")
+    checkAnswer(df4, Seq(Row("abc_%def@gmail.com")))
+
+    val df5 = spark.table("h2.test.address").filter($"email".startsWith("abc_'%"))
+    checkFiltersRemoved(df5)
+    checkPushedInfo(df5, "PushedFilters: [EMAIL IS NOT NULL, EMAIL LIKE 'abc\\_'\\%%']")
```

Comment:
interesting, is this a valid SQL? how does the parser parse `EMAIL LIKE 'abc\\_'\\%%'`?

Reply:
I don't know the background, but it works good.

Reply:
At least snowflake asks you to quote it: https://docs.snowflake.com/en/sql-reference/data-types-text#escape-sequences-in-single-quoted-string-constants

Are you sure this works for all databases not only H2?

Reply:
I'm not sure. I guess it is different from each databases.

Reply:
then let's quote it like other special chars to be safe.

Reply:
In fact, DS V2 pass `"EMAIL" LIKE 'abc\_''\%%'` to H2.